File: ContestOfSharp/ContestOfSharp/Challenges/Challenge1.cs
/*
 * Escribe un programa que muestre por consola (con un print) los
 * números de 1 a 100 (ambos incluidos y con un salto de línea entre
 * cada impresión), sustituyendo los siguientes:
 * - Múltiplos de 3 por la palabra "fizz".
 * - Múltiplos de 5 por la palabra "buzz".
 * - Múltiplos de 3 y de 5 a la vez por la palabra "fizzbuzz".
 */

using static System.Linq.Enumerable;
using static BankaiSouth.LanguageExtensions;
using BankaiSouth;

namespace ContestOfSharp.Challenges;

public abstract class Challenge1
{
    private const string MultiplesOfThreeReplacement = "fizz";
    private const string MultiplesOfFiveReplacement = "buzz";
    private const string MultiplesOfThreeAndFiveReplacement = "fizzbuzz";

    private static string FizzBuzzImperative()
    {
        var toPrint = "";
        
        foreach (var n in Range(1, 100))
        {
            if (n > 1) toPrint += "\n";

            if (n.IsMultipleOf(number: 3) && n.IsMultipleOf(number: 5))
            {
                toPrint += MultiplesOfThreeAndFiveReplacement;
            }
            else if (n.IsMultipleOf(number: 3))
            {
                toPrint += MultiplesOfThreeReplacement;
            }
            else if (n.IsMultipleOf(number: 5))
            {
                toPrint += MultiplesOfFiveReplacement;
            }
            else
            {
                toPrint += n;
            }
        }

        return toPrint;
    }

    private static string FizzBuzzDeclarative()
        => Range(1, 100)
            .Aggregate(Array.Empty<string>(), (strings, n) =>
            {
                return When(rules: new (bool, string[])[]
                {
                    (n.IsMultipleOf(number: 3) && n.IsMultipleOf(number: 5),
                        strings.Adding(element: MultiplesOfThreeAndFiveReplacement)),
                    (n.IsMultipleOf(number: 3),
                        strings.Adding(element: MultiplesOfThreeReplacement)),
                    (n.IsMultipleOf(number: 5),
                        strings.Adding(element: MultiplesOfFiveReplacement))
                }, fallback: strings.Adding(element: n.ToString()));
            })
            .Joined(separator: "\n");

    internal static void Perform()
    {
        Console.WriteLine("FizzBuzz Imperative ------------------------------------------");
        Console.WriteLine(FizzBuzzImperative());
        Console.WriteLine("FizzBuzz Declarative ------------------------------------------");
        Console.WriteLine(FizzBuzzDeclarative());
    }
}

File: contestOfKotlin/src/main/kotlin/challenges/Challenge1.kt
package org.example.challenges.challenge1

/*
 * Escribe un programa que muestre por consola (con un print) los
 * números de 1 a 100 (ambos incluidos y con un salto de línea entre
 * cada impresión), sustituyendo los siguientes:
 * - Múltiplos de 3 por la palabra "fizz".
 * - Múltiplos de 5 por la palabra "buzz".
 * - Múltiplos de 3 y de 5 a la vez por la palabra "fizzbuzz".
 */

private const val MULTIPLES_OF_THREE_REPLACEMENT = "fizz"
private const val MULTIPLES_OF_FIVE_REPLACEMENT = "buzz"
private const val MULTIPLES_OF_THREE_AND_FIVE_REPLACEMENT = "fizzbuzz"
fun fizzBuzz(): String {
    var toPrint = ""

    for (n in 1..100) {
        if (n > 1) {
            toPrint += "\n"
        }

        toPrint += when {
            isNumber(n, multipleOf = 3) && isNumber(n, multipleOf = 5) -> MULTIPLES_OF_THREE_AND_FIVE_REPLACEMENT
            isNumber(n, multipleOf = 3) -> MULTIPLES_OF_THREE_REPLACEMENT
            isNumber(n, multipleOf = 5) -> MULTIPLES_OF_FIVE_REPLACEMENT
            else -> n.toString()
        }
    }

    return toPrint
}
fun isNumber(number: Int, multipleOf: Int): Boolean {
    return number % multipleOf == 0
}

fun performFizzBuzz() {
    println(fizzBuzz())
}